realm/realm-library/src/main/cpp/io_realm_internal_objectstore_OsSubscriptionSet.cpp

```diff
@@ -0,0 +1,189 @@
+{
+    try {
+        auto subscriptions = reinterpret_cast<sync::SubscriptionSet*>(j_subscription_set_ptr);
+        sync::SubscriptionSet::State state = subscriptions->state();
+        switch(state) {
+            case sync::SubscriptionSet::State::Uncommitted:
+                return io_realm_internal_objectstore_OsSubscriptionSet_STATE_VALUE_UNCOMMITTED;
+            case sync::SubscriptionSet::State::Pending:
+                return io_realm_internal_objectstore_OsSubscriptionSet_STATE_VALUE_PENDING;
+            case sync::SubscriptionSet::State::Bootstrapping:
+                return io_realm_internal_objectstore_OsSubscriptionSet_STATE_VALUE_BOOTSTRAPPING;
+            case sync::SubscriptionSet::State::Complete:
+                return io_realm_internal_objectstore_OsSubscriptionSet_STATE_VALUE_COMPLETE;
+            case sync::SubscriptionSet::State::Error:
+                return io_realm_internal_objectstore_OsSubscriptionSet_STATE_VALUE_ERROR;
+            case sync::SubscriptionSet::State::Superceded:
+                return io_realm_internal_objectstore_OsSubscriptionSet_STATE_VALUE_SUPERCEDED;
+        }
+    }
+    CATCH_STD()
+    return 0;
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_objectstore_OsSubscriptionSet_nativeCreateMutableSubscriptionSet(JNIEnv* env, jclass, jlong j_subscription_set_ptr)
+{
+    try {
+        auto subscriptions = reinterpret_cast<sync::SubscriptionSet*>(j_subscription_set_ptr);
+        return reinterpret_cast<jlong>(new sync::MutableSubscriptionSet(subscriptions->make_mutable_copy()));
+    }
+    CATCH_STD()
+    return 0;
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_objectstore_OsSubscriptionSet_nativeSubscriptionAt(JNIEnv *env, jclass,
+                                                                          jlong j_subscription_set_ptr,
+                                                                          jint j_index)
+{
+    try {
+        auto subscriptions = reinterpret_cast<sync::SubscriptionSet*>(j_subscription_set_ptr);
+        return reinterpret_cast<jlong>(new sync::Subscription(subscriptions->at(j_index)));
+    }
+    CATCH_STD()
+    return 0;
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_objectstore_OsSubscriptionSet_nativeWaitForSynchronization(JNIEnv *env,
+                                                                                  jclass,
+                                                                                  jlong j_subscription_set_ptr,
+                                                                                  jobject j_callback)
+{
+    try {
+        auto subscriptions = reinterpret_cast<sync::SubscriptionSet *>(j_subscription_set_ptr);
+        util::Future<sync::SubscriptionSet::State> result = subscriptions->get_state_change_notification(
+                sync::SubscriptionSet::State::Complete);
+
+        static JavaClass callback_class(env, "io/realm/internal/objectstore/OsSubscriptionSet$StateChangeCallback");
```

Comment:
Are we protecting `OsSubscriptionSet$StateChangeCallback` from proguard?

Reply:
No, good catch